[SPARK-40646][SQL] Fix returning partial results in JSON data source and JSON functions

sadikovi opened the pull request:
<!--
Thanks for sending a pull request!  Here are some tips for you:
  1. If this is your first time, please read our contributor guidelines: https://spark.apache.org/contributing.html
  2. Ensure you have added or run the appropriate tests for your PR: https://spark.apache.org/developer-tools.html
  3. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP][SPARK-XXXX] Your PR title ...'.
  4. Be sure to keep the PR description updated to reflect all changes.
  5. Please write your PR title to summarize what this PR proposes.
  6. If possible, provide a concise example to reproduce the issue for a faster review.
  7. If you want to add a new configuration, please read the guideline first for naming configurations in
     'core/src/main/scala/org/apache/spark/internal/config/ConfigEntry.scala'.
  8. If you want to add or modify an error type or message, please read the guideline first in
     'core/src/main/resources/error/README.md'.
-->

### What changes were proposed in this pull request?
<!--
Please clarify what changes you are proposing. The purpose of this section is to outline the changes and how this PR fixes the issue. 
If possible, please consider writing useful notes for better and faster reviews in your PR. See the examples below.
  1. If you refactor some codes with changing classes, showing the class hierarchy will help reviewers.
  2. If you fix some SQL features, you can provide some references of other DBMSes.
  3. If there is design documentation, please add the link.
  4. If there is a discussion in the mailing list, please add the link.
-->

This PR is a follow-up for [SPARK-33134](https://issues.apache.org/jira/browse/SPARK-33134) (https://github.com/apache/spark/pull/30031). 

I found another case when, depending on the order of columns, parsing one JSON field breaks all of the subsequent fields resulting in all nulls:

With a file like this:
```
{"a": {"x": 1, "y": true}, "b": {"x": 1}}
{"a": {"x": 2}, "b": {"x": 2}}
```

Reading the file results in column `b` as null even though it is a valid column.
```scala
val df = spark.read
  .schema("a struct<x: int, y: struct<x: int>>, b struct<x: int>")
  .json("path") 

===

a	                b
null	                null
{"x":2,"y":null}	{"x":2} 
```

However, b column should be:
```
{"x": 1}
{"x": 2}
```

This particular example actually used to work in earlier Spark versions but it was affected by SPARK-33134 which fixed another bug with the incorrect parsing in `from_json`. Because this case was not tested, we missed it at the time. 

In order to fix both SPARK-33134 and SPARK-40646, we need to process `PartialResultException` in `convertArray` method to handle any errors in child objects. Without the fix, the code would not wrap the row in the array for `from_json` resulting in a ClassCastException (SPARK-33134). Because of this handling, we don't need `isRoot` check anymore in `convertObject` thus unblocking SPARK-40646.

I updated the code to handle both cases. With these changes, we can correctly parse this case:
```scala
val df3 = Seq("""[{"c2": [19], "c1": 123456}]""").toDF("c0")
checkAnswer(df3.select(from_json($"c0", ArrayType(st))), Row(Array(Row(123456, null))))
```
which was previously returning `null` for the root row.

### Why are the changes needed?
<!--
Please clarify why the changes are needed. For instance,
  1. If you propose a new API, clarify the use case for a new API.
  2. If you fix a bug, you can clarify why it is a bug.
-->

Fixes a long-standing issue when parsing a JSON with an incorrect field that would break parsing of the entire record.

### Does this PR introduce _any_ user-facing change?
<!--
Note that it means *any* user-facing change including all aspects such as the documentation fix.
If yes, please clarify the previous behavior and the change this PR proposes - provide the console output, description and/or an example to show the behavior difference if possible.
If possible, please also clarify if this is a user-facing change compared to the released Spark versions or within the unreleased branches such as master.
If no, write 'No'.
-->

No.

### How was this patch tested?
<!--
If tests were added, say they were added here. Please make sure to add some test cases that check the changes thoroughly including negative and positive cases if possible.
If it was tested in a way different from regular unit tests, please clarify how you tested step by step, ideally copy and paste-able, so that other reviewers can test and check, and descendants can verify in the future.
If tests were not added, please describe why they were not added and/or why it was difficult to add.
If benchmark tests were added, please run the benchmarks in GitHub Actions for the consistent environment, and the instructions could accord to: https://spark.apache.org/developer-tools.html#github-workflow-benchmarks.
-->

I added unit tests for SPARK-40646 as well as SPARK-33134.

Commit:
update code

sadikovi commented:
cc @MaxGekk @HyukjinKwon, I would appreciate a review. Let me know if you have any concerns with the changes. Thank you.

Commit:
fix scalastyle

MaxGekk commented:
My related PR https://github.com/apache/spark/pull/23253

AmplabJenkins commented:
Can one of the admins verify this patch?

Commit:
Merge remote-tracking branch 'upstream/master' into SPARK-40646

MaxGekk commented on sql/core/src/test/scala/org/apache/spark/sql/JsonFunctionsSuite.scala:
```diff
@@ -853,11 +853,68 @@ class JsonFunctionsSuite extends QueryTest with SharedSparkSession {
+      .add("c1", StringType)
+      .add("c2", ArrayType(new StructType().add("a", LongType)))
+
+    // Value of "c2.a" is a string instead of a long.
```
Could you explain this comment `"c2.a" is a string`, and doesn't it contradict to `"c2": {"a": 1}`. Is `1` string?

sadikovi replied:
Thanks for pointing it out, the comment is incorrect, it belonged to the original test that I refactored later. I will update.

MaxGekk commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonParser.scala:
```diff
@@ -456,7 +456,7 @@ class JacksonParser(
           } catch {
             case e: SparkUpgradeException => throw e
-            case NonFatal(e) if isRoot =>
+            case NonFatal(e) =>
```
If you allow this, please, check all complex types `map`, `struct`. Or have you added some tests already?

sadikovi replied:
I have tried covering different cases in:
- SPARK-40646: return partial results for JSON arrays with objects
- SPARK-40646: return partial results for objects with values as JSON arrays

I think I am missing a test for map, will add, thanks.

sadikovi replied:
I added a test!

Commit:
handle maps

sadikovi commented:
@MaxGekk I have addressed your comments, could you review again?

I think there is still an issue with complex fields:
```scala
test("SPARK-40646") {
  val st = new StructType()
    .add("c1", MapType(StringType, new StructType().add("a1", LongType)))
    .add("c2", StringType)

  // Map "c2" has "k2" key that is a string, not an integer.
  val df = Seq("""{"c1": {"k1": {"a1": 1}, "k2": {"a1": "A"}, "k3": {"a1": 2}}}, "c2": "abc"}""").toDF("c0")
  checkAnswer(
    df.select(from_json($"c0", st)),
    Row(Row(null, "abc"))
  )
}
```

would return `null, null` instead of `null, "abc"`. 

Would it be okay to address that case in the follow-up? It might require a bit of time to fix that case.

sadikovi commented:
cc @cloud-fan

MaxGekk commented:
> Would it be okay to address that case in the follow-up?

Sure.

MaxGekk commented:
+1, LGTM. Merging to master.
Thank you, @sadikovi.

sadikovi commented:
Thanks for merging!
